### PR TITLE
Enable using the write ahead log

### PR DESCRIPTION
**Pull request by wi11dey**

### CODE REVIEW

File: src/java/org/apache/cassandra/db/compaction/CompactionTask.java [hunk 1]
```diff
@@ -156,122 +156,127 @@ public boolean apply(SSTableReader sstable)
...
-                logger.error("Failed to write to the write-ahead log for {}/{}. Continuing to compact now can cause resurrection. Exiting",
-                             cfs.keyspace.getName(), cfs.name, e);
-                System.exit(1);
+        }
+        finally
+        {
+            if (!readyToFinish) {
+                // TODO(wdey): refactor all of the trys
+                try (Refs closedRefs = refs; CompactionController closedController = controller) {}
             }
```
Comment (wi11dey) [inline src/java/org/apache/cassandra/db/compaction/CompactionTask.java:235]: same comment as https://github.com/palantir/cassandra/pull/557#discussion_r1809277087

**Comment by rhuffy** [comment]
Do you want to use safelogging here?

**Comment by wi11dey** [comment]
an indentation level was removed in the outer try, no logging changes in this pr; best to click "Hide whitespace" for the diff

File: src/java/org/apache/cassandra/db/compaction/CompactionTask.java [hunk 2]
```diff
@@ -157,122 +158,128 @@ public boolean apply(SSTableReader sstable)
...
+        catch (Exception e)
+        {
+            logger.error("Failed to write to the write-ahead log for {}/{}. Continuing to compact now can cause resurrection. Exiting",
+                         cfs.keyspace.getName(), cfs.name, e);
+            panic();
+        }
+        refs.close();
+        controller.close();
+        if (taskId != null)
+            SystemKeyspace.finishCompaction(taskId);
```
Comment (zpear) [inline src/java/org/apache/cassandra/db/compaction/CompactionTask.java:259]: Do we have any tests that assert if any ancestors of a compaction are missing, the products are all left alone?

File: src/java/org/apache/cassandra/db/compaction/CompactionTask.java [hunk 3]
```diff
@@ -157,122 +158,128 @@ public boolean apply(SSTableReader sstable)
...
-                cfs.metric.compactionsCompleted.inc();
-            }
+        try
+        {
+            ColumnFamilyStoreManager.instance.markForDeletion(cfs.metadata, transaction.logged.obsoleteDescriptors());
+        }
+        catch (Exception e)
+        {
+            logger.error("Failed to write to the write-ahead log for {}/{}. Continuing to compact now can cause resurrection. Exiting",
+                         cfs.keyspace.getName(), cfs.name, e);
```
Comment (zpear) [inline src/java/org/apache/cassandra/db/compaction/CompactionTask.java]: nit: can you safearg the keyspace + cf in the panic error logs

File: src/java/org/apache/cassandra/db/compaction/CompactionTask.java (hunk 2 above)
Comment (wi11dey) [reply src/java/org/apache/cassandra/db/compaction/CompactionTask.java:259]: added testRemoveUnusedSstablesDoesNotTouchProductsWhenAncestorDoesNotExist to ColumnFamilyStoreTest

**Review by zpear — APPROVED** [review]

**Pull request merged**